File: backend/app/agents/manager_agent.py
from typing import List
from app.models.schemas import AgentReport, ManagerReport

def run(reports: List[AgentReport]) -> ManagerReport:
    recs = {r.agent_name: r.recommendation for r in reports}
    scores = {r.agent_name: r.score for r in reports}
    disagreements=[]
    if len(set(recs.values())) > 1:
        disagreements.append(f"Agents disagree: {recs}")
    if max(scores.values()) - min(scores.values()) > 0.35:
        disagreements.append("Large confidence gap between specialist scores.")
    requested=[]
    for r in reports:
        if r.recommendation == "REFER" and r.score < 0.55:
            requested.append(f"Request targeted re-analysis from {r.agent_name}")
    if any(r.recommendation == "REJECT" for r in reports):
        recommendation="REFER"
    elif disagreements or any(r.recommendation == "REFER" for r in reports):
        recommendation="REFER"
    else:
        recommendation="APPROVE"
    all_reasons=[]
    for r in reports:
        all_reasons.extend(r.reason_codes)
    reviewer_summary = " | ".join([f"{r.agent_name}: {r.recommendation} ({r.score})" for r in reports])
    explanation = "Manager summarizes specialist outputs only. It does not change scores or approve credit. Main reason codes: " + ", ".join(sorted(set(all_reasons)))
    return ManagerReport(recommendation=recommendation, disagreements=disagreements, requested_reanalysis=requested, reviewer_summary=reviewer_summary, readable_explanation=explanation)

File: frontend/lib/backend/agents/manager-agent.ts
import type { AgentReport, ManagerReport } from "../schemas";

export function runManagerAgent(reports: AgentReport[]): ManagerReport {
  const recs = Object.fromEntries(
    reports.map((r) => [r.agent_name, r.recommendation])
  );
  const scores = Object.fromEntries(
    reports.map((r) => [r.agent_name, r.score])
  );
  const disagreements: string[] = [];

  if (new Set(Object.values(recs)).size > 1) {
    disagreements.push(`Agents disagree: ${JSON.stringify(recs)}`);
  }

  const scoreValues = Object.values(scores);
  if (Math.max(...scoreValues) - Math.min(...scoreValues) > 0.35) {
    disagreements.push("Large confidence gap between specialist scores.");
  }

  const requested: string[] = [];
  for (const r of reports) {
    if (r.recommendation === "REFER" && r.score < 0.55) {
      requested.push(`Request targeted re-analysis from ${r.agent_name}`);
    }
  }

  let recommendation: string;
  if (reports.some((r) => r.recommendation === "REJECT")) {
    recommendation = "REFER";
  } else if (
    disagreements.length ||
    reports.some((r) => r.recommendation === "REFER")
  ) {
    recommendation = "REFER";
  } else {
    recommendation = "APPROVE";
  }

  const allReasons = reports.flatMap((r) => r.reason_codes);
  const reviewerSummary = reports
    .map((r) => `${r.agent_name}: ${r.recommendation} (${r.score})`)
    .join(" | ");

  return {
    recommendation,
    disagreements,
    requested_reanalysis: requested,
    reviewer_summary: reviewerSummary,
    readable_explanation:
      "Manager summarizes specialist outputs only. It does not change scores or approve credit. Main reason codes: " +
      [...new Set(allReasons)].sort().join(", "),
  };
}
